### ISSUE TITLE
CIFAR 10 training example backward pass

### ISSUE THREAD

**Issue by calvinqi** [Tiiiger/QPyTorch#19]
Hi again! Just wondering, in the CIFAR10_Low_Precision_Training_Example notebook, why is the backward_number for weights, gradients, and momentum set to None, while the activations/errors have low precision backward passes. Shouldn't they all be in low precision mode to be trained as according to the paper?

Maybe I am misunderstanding the roles of the different parameters.
Thank you!

**Comment by Tiiiger** [comment]
Hi, 

apologies for the confusion. The quantizations for weights, gradients, and momentum only involve a forward pass. Take weight quantization for an example. The quantization is done outside of the normal pytorch scope of auto-differentiation, manually in the low precision optimizer. 

Note that this does not mean that there's a full-precision backward pass for the weight quantization; there simply isn't any backward pass. This is more an implementation choice rather than an actual difference in the algorithm.

I am not sure how best to explain it. But if you feel comfortable reading at the source code, you can see [here](https://github.com/Tiiiger/QPyTorch/blob/master/qtorch/optim/optim_low.py#L92) that when we are quantizing the weights, we only need to call the forward pass. 

I feel like I just did a horrible job explaining. Please follow up with more clarification questions and I think I can do better explain more specific things. I think it would be helpful if you can point me to the paper you're looking at (you said _according to the paper_) and I can use that as a context.

**Comment by Tiiiger** [comment]
I am closing this. Feel free to reopen if you have any further questions.

**Comment by calvinqi** [comment]
Hello! Apologies for the delay. Want to make sure I'm understanding this correctly:
* The normal gradient is computed using autodiff in the background
* OptimLP quantizes the gradient separate from autograd
* The `optim.step()` is performed with the quantized gradient, affecting the weight
* The weight is quantized after

As a follow up question, how come `act_error_quant` has a backward number? How is that one computed and why is it different?
